### PR TITLE
Dataset.map

### PR DESCRIPTION
**Pull request by max-sixty**
- [x] Helps https://github.com/pydata/xarray/issues/1251
 - [x] Tests added
 - [x] Passes `black . && mypy . && flake8`
 - [x] Fully documented, including `whats-new.rst` for all changes and `api.rst` for new API

This is the first step towards organizing these functions a bit better, as outlined in https://github.com/pydata/xarray/issues/1251, https://github.com/pydata/xarray/issues/1618, https://github.com/pydata/xarray/pull/2674

While this one is a fairly easy decision, others are going to be harder. Open to a more methodical up-front process before we start making changes, if we think that's necessary.

### CODE REVIEW

**Comment by max-sixty** [comment]
I'm getting bemuisng test failures here! All the tests apart from docs say they pass, Docs says `Cancelled after 60m`. There are no explicit changes to docs here, and I can't imagine any implicit changes. If anyone has any bright ideas, help is welcome!

**Comment by max-sixty** [comment]
I see this in the [GH Checks page](https://github.com/pydata/xarray/runs/280352750) (but, again, no test runs failed, apparently)


```python
self = <xarray.tests.test_backends.TestPydapOnline object at 0x7fd9047f8518>

    def test_cmp_local_file(self):
>       with self.create_datasets() as (actual, expected):

xarray/tests/test_backends.py:3137: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 
/usr/share/miniconda/envs/xarray-tests/lib/python3.6/contextlib.py:81: in __enter__
    return next(self.gen)
xarray/tests/test_backends.py:3191: in create_datasets
    actual = open_dataset(url, engine="pydap", **kwargs)
xarray/backends/api.py:504: in open_dataset
    store = backends.PydapDataStore.open(filename_or_obj, **backend_kwargs)
xarray/backends/pydap_.py:78: in open
    ds = pydap.client.open_url(url, session=session)
/usr/share/miniconda/envs/xarray-tests/lib/python3.6/site-packages/pydap/client.py:67: in open_url
    timeout).dataset
/usr/share/miniconda/envs/xarray-tests/lib/python3.6/site-packages/pydap/handlers/dap.py:54: in __init__
    raise_for_status(r)
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 

response = <Response at 0x7fd9045ac588 500 Internal Server Error>

    def raise_for_status(response):
        # Raise error if status is above 300:
        if response.status_code >= 300:
            raise HTTPError(
                detail=response.status+'\n'+response.text,
                headers=response.headers,
>               comment=response.body
            )
E           webob.exc.HTTPError: 500 Internal Server Error
E           Error { 
E               code = 500;
E               message = "Problem encountered with BES connection. Message: 'Failed to write to socket: Broken pipe (Write failed)' OPeNDAPClient executed 131 prior commands.";
E           }

/usr/share/miniconda/envs/xarray-tests/lib/python3.6/site-packages/pydap/net.py:39: HTTPError
```

**Comment by max-sixty** [comment]
I think docs stopped building and just froze (so timed out rather than erroring), and so the error wasn't reported up to Azure Pipelines. I got this from the log:

```
>>>-------------------------------------------------------------------------
Warning in /home/vsts/work/1/s/doc/computation.rst at block ending on line 472
Specify :okwarning: as an option in the ipython:: block to suppress this message
----------------------------------------------------------------------------
/usr/share/miniconda/envs/xarray-tests/bin/sphinx-build:1: DeprecationWarning: Dataset.apply is deprecated in favor of Dataset.map and will be changed or removed in a future version of xarray
  #!/usr/share/miniconda/envs/xarray-tests/bin/python3.7
<<<-------------------------------------------------------------------------
/home/vsts/work/1/s/xarray/core/dataarray.py:docstring of xarray.DataArray.integrate:12: WARNING: Unexpected indentation.
/home/vsts/work/1/s/xarray/core/dataarray.py:docstring of xarray.DataArray.interp:20: WARNING: Inline strong start-string without end-string.
/home/vsts/work/1/s/xarray/core/dataarray.py:docstring of xarray.DataArray.interpolate_na:8: WARNING: Definition list ends without a blank line; unexpected unindent.

Sphinx parallel build error:
RuntimeError: Non Expected warning in `/home/vsts/work/1/s/doc/computation.rst` line 472
```

**Comment by max-sixty** [comment]
This is ready for review & merge

**Comment by max-sixty** [comment]
Ready to merge!

File: xarray/core/dataset.py [hunk 1]
```diff
@@ -4165,6 +4165,20 @@ def apply(
         attrs = self.attrs if keep_attrs else None
         return type(self)(variables, attrs=attrs)
 
+    def apply(
+        self,
+        func: Callable,
+        keep_attrs: bool = None,
+        args: Iterable[Any] = (),
+        **kwargs: Any,
+    ) -> "Dataset":
```
Comment (shoyer) [inline xarray/core/dataset.py]: Can we keep the docstring here, even if it's just "Alias for Dataset.map"?

Ideally we would use the `See also` section, which gets turned into a link by numpydoc.

Comment (max-sixty) [reply xarray/core/dataset.py]: Yes for sure, that was an oversight (e.g. I kept them as references for `drop`)

Comment (max-sixty) [reply xarray/core/dataset.py]: Also added `See Also` for the drop cases

Comment (max-sixty) [reply xarray/core/dataset.py]: Yes for sure

**Comment by max-sixty** [comment]
Lmk if that was an "OK to merge" @shoyer ! Thanks

**Comment by shoyer** [comment]
Looks good to me!

File: doc/whats-new.rst [hunk 2]
```diff
@@ -44,6 +44,13 @@ New Features
   option for dropping either labels or variables, but using the more specific methods is encouraged.
   (:pull:`3475`)
   By `Maximilian Roos <https://github.com/max-sixty>`_
+- :py:meth:`Dataset.map` & :py:meth:`GroupBy.map` & :py:meth:`Resample.map` have been added for
```
Comment (dcherian) [inline doc/whats-new.rst]: I think we need these in api.rst? or is it automatically added?

Comment (max-sixty) [reply doc/whats-new.rst]: Ah, thanks @dcherian , good spot

**Pull request merged**